### PR TITLE
Feat(rln-relay): enables two modes of onchain and offchain

### PR DESCRIPTION
**Pull request by staheri14**
This PR is a significant step towards completing https://github.com/status-im/nwaku/issues/894 

**Summary of changes:**
- Splits the rln relay mounting procedure into two separate smaller procs, one to mount rln-relay with static and off-chain group management and the other for the dynamic and on-chain group management.  Thus, the protocol can be mounted in two different modes. A proper config option is also introduced to distinguish these modes.  
- It also integrates the on-chain/dynamic mode of the rln-relay protocol into the waku node, as such nodes running rln-relay in on-chain/dynamic mode are able to listen to the contract and dynamically update their membership tree locally. Proper functional and integration tests are also included in this PR.
- Introduces some config options related to the on-chain/dynamic mode of the rln-relay protocol.
- The developed `mountRlnRelayDynamic` proc additionally and optionally takes care of the registration of the peers without rln-relay credentials into the membership contract. The corresponding config options are incorporated to get rln credentials from the user if available or otherwise let the `mountRlnRelayDynamic` take care of credential creation and registration. 
- Implements a custom solution to extract (decode and convert) the event arguments i.e., the rln identity commitment key and their index from the bytes of the transaction receipt (the missing functionality in the nim-web3 library). This is necessary as it allows to identify the index of the registered member's identity commitment key in the Merkle tree and hence the corresponding peer would be able to properly generate proofs.

## Next steps
- Exploring the use of Keystore to persist rln credentials + eth credentials (or other alternative low-hanging solutions)
- Conducting a test on a local network of rln-relay peers all running in the on-chain mode to identify any remaining issues.
- Integrating the dynamic rln-relay into chat2
- Preparing a tutorial on testing on-chain rln-relay using chat2 and Ganache as the Blockchain simulator.

### CODE REVIEW

File: waku/v2/node/config.nim [hunk 1]
```diff
@@ -129,6 +130,34 @@ type
       defaultValue: "/toy-chat/2/luzhou/proto"
       name: "rln-relay-content-topic" }: ContentTopic
     
+    rlnRelayDynamic* {.
```
Comment (jm-clius) [inline waku/v2/node/config.nim:133]: Perhaps not yet for this PR, but a good case for why we need to start using sub-commands as explained in https://github.com/status-im/nim-confutils to group related commands together :)

**Review by jm-clius — COMMENTED** [review]
Thanks for this! I'll be reviewing more in-depth as well. For now my main comment relates to separating RLN logic structurally in the code (especially moving the `mount`-procs and setup code to a separate module). I think in future we want to move the `mount`-logic for all protocols to the respective protocol modules in any case.

File: waku/v2/node/wakunode2.nim [hunk 2]
```diff
@@ -1201,29 +1220,51 @@ when isMainModule:
     
     when defined(rln): 
       if conf.rlnRelay:
-        info "WakuRLNRelay is enabled"
-
-        # set up rln relay inputs
-        let (groupOpt, memKeyPairOpt, memIndexOpt) = rlnRelaySetUp(conf.rlnRelayMemIndex)
-        if memIndexOpt.isNone:
-          error "failed to mount WakuRLNRelay"
+        if not conf.rlnRelayDynamic:
```
[the remark this answers is not in the document]
Comment (staheri14) [reply waku/v2/node/wakunode2.nim:1223]: Thanks for the suggestion, fair enough, here is the tracking issue https://github.com/status-im/nwaku/issues/998 which I will address in a separate PR

File: waku/v2/node/config.nim (hunk 1 above)
Comment (staheri14) [reply waku/v2/node/config.nim:133]: Sure, sounds a good idea, much cleaner and organized! here is the tracking issue https://github.com/status-im/nwaku/issues/999

**Comment by staheri14** [comment]
> Thanks for this! I'll be reviewing more in-depth as well. For now my main comment relates to separating RLN logic structurally in the code (especially moving the `mount`-procs and setup code to a separate module). I think in future we want to move the `mount`-logic for all protocols to the respective protocol modules in any case.

Thanks for your comment, I agree, and I will address it in a separate PR (I am already on it) to let this PR enjoy a more defined and limited scope of changes.

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 3]
```diff
@@ -99,28 +101,65 @@ proc membershipKeyGen*(ctxPtr: RLN[Bn256]): Option[MembershipKeyPair] =
...
-  discard await sender.register(pk).send(MEMBERSHIP_FEE)
-  debug "pk", pk = pk
-  # TODO check the receipt and then return true/false
+  # web3.privateKey = some(ethAccountPrivateKey)
+  var sender = web3.contractSender(MembershipContract, membershipContractAddress) # creates a Sender object with a web3 field and contract address of type Address
+
+  debug "registering an id commitment", idComm=idComm
+  let pk = idComm.toUInt256()
+  let txHash = await sender.register(pk).send(MEMBERSHIP_FEE)
+  let tsReceipt = await web3.getMinedTransactionReceipt(txHash)
```
[the remark this answers is not in the document]
Comment (staheri14) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim]: done https://github.com/status-im/nwaku/pull/992/commits/4371876b36573a9421ca6b759616f09f84c135d8

**Comment by staheri14** [comment]
@jm-clius @LNSD Please let me know your comments on the PR.

**Review by jm-clius — APPROVED** [review]
LGTM! Sorry for the long wait for the review :). Would be good to see the structural/module split in a subsequent PR to keep wakunode2 a bit more manageable.

**Comment by staheri14** [comment]
@LNSD just a gentle reminder, thanks!

**Pull request merged**